### PR TITLE
Dockerise and add CI pipeline to Gem

### PR DESCRIPTION
**Pull request by banjoko01**
This PR does the following:
- Dockerize the gem with support for `Ruby 2.7`
- Add Github CI pipeline workflow to gem

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  release:
+    needs: test
+    if: startsWith(github.ref, 'refs/tags/v') || startsWith(github.ref, 'refs/tags/build-')
+    runs-on: ubuntu-latest
+
+    env:
+      GEMS_PATH: "pkg/*.gem"
+      RUBYGEMS_HOST: "https://sageonegems.jfrog.io/artifactory/api/gems/gems-local"
+      JFROG_USER: "${{ secrets.JFROG_USER }}"
+      JFROG_PASS: "${{ secrets.JFROG_PASS }}"
```
Comment (ianoxley) [inline .github/workflows/ci.yml:42]: Is this correct for a public repo?

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+version: '2'
+
+services:
+  client:
+    build:
+      context: .
+      dockerfile: Dockerfile
+      args:
+        - BUNDLE_SAGEONEGEMS__JFROG__IO
```
Comment (ianoxley) [inline docker-compose.yml:9]: Is this needed? It's a public repo using rubygems.org as the gem source...

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    image: sageone/zuora
+    container_name: zuora
+    environment:
+     - BUNDLE_SAGEONEGEMS__JFROG__IO
```
Comment (ianoxley) [inline docker-compose.yml:13]: Is this needed? It's a public repo using rubygems.org as the gem source...

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,25 @@
+FROM ruby:2.7-alpine
```
Comment (ianoxley) [inline Dockerfile:1]: We can use an arg for the Ruby version here:
```suggestion
ARG RUBY_VERSION=${RUBY_VERSION:-3.2}
FROM ruby:${RUBY_VERSION}-alpine
```

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+version: '2'
+
+services:
+  client:
+    build:
+      context: .
+      dockerfile: Dockerfile
```
Comment (ianoxley) [inline docker-compose.yml:7]: Then add the build args in here:
```suggestion
      dockerfile: Dockerfile
      args:
        - RUBY_VERSION
```
We should be able to boot it with different ruby version via: `RUBY_VERSION=2.7 docker-compose up -d --build`

File: ZUORA_VERSION [hunk 6]
```diff
@@ -0,0 +1 @@
+0.0.2-sage
```
Comment (ianoxley) [inline ZUORA_VERSION:1]: Is this file needed?

File: spec/integration/account_management_spec.rb [hunk 7]
```diff
@@ -1,6 +1,6 @@
 require 'spec_helper'
 
-describe "Account Management" do
+describe "Account Management", :skip do
```
Comment (ianoxley) [inline spec/integration/account_management_spec.rb:3]: Why are we skipping these tests?

File: spec/integration/product_catalog_spec.rb [hunk 8]
```diff
@@ -1,6 +1,6 @@
 require 'spec_helper'
 
-describe "Product Catalog" do
+describe "Product Catalog", :skip do
```
Comment (ianoxley) [inline spec/integration/product_catalog_spec.rb:3]: Why are we skipping these tests?

File: spec/integration/subscription_spec.rb [hunk 9]
```diff
@@ -1,55 +1,52 @@
 require 'spec_helper'
 
-describe "Subscription" do
-
+describe 'Subscription', :skip do
```
Comment (ianoxley) [inline spec/integration/subscription_spec.rb:3]: Why are we skipping these tests?

File: spec/zuora/api_spec.rb [hunk 10]
```diff
@@ -21,8 +21,8 @@
       Zuora::Api.instance.config.password.should == 'changed'
     end
   end
-  
-  describe "logger support" do
+
+  describe "logger support", :skip do
```
Comment (ianoxley) [inline spec/zuora/api_spec.rb:25]: Why `:skip`?

File: spec/zuora/api_spec.rb [hunk 11]
```diff
@@ -32,7 +32,7 @@
     end
   end
 
-  describe "authentication" do
+  describe "authentication", :skip do
```
Comment (ianoxley) [inline spec/zuora/api_spec.rb:35]: Why `:skip`?

File: spec/zuora/api_spec.rb [hunk 12]
```diff
@@ -59,7 +59,7 @@
     end
   end
 
-  describe :download do
+  describe 'download', :skip do
```
Comment (ianoxley) [inline spec/zuora/api_spec.rb:62]: Why `:skip`?

File: zuora.gemspec [hunk 13]
```diff
@@ -1,36 +1,34 @@
...
+  s.add_dependency 'activemodel', '< 5.0.0'
+  s.add_dependency 'activesupport', '< 5.0.0'
+  s.add_dependency 'akami', '1.3.2'
+  s.add_dependency 'ffi', '1.16.3'
+  s.add_dependency 'httpi', '~> 1.0'
+  s.add_dependency 'libxml4r', '~> 0.2.6'
+  s.add_dependency 'nokogiri', '~> 1.15.6'
+  s.add_dependency 'ruby3-backward-compatibility'
+  s.add_dependency 'savon', '~> 0.9.12'
+  s.add_dependency 'wasabi', '2.3.0'
```
Comment (ianoxley) [inline zuora.gemspec:33]: We seem to have doubled the non-development dependencies from 5 to 10. Are all these changes necessary?

`ruby3-backward-compatibility` definitely isn't needed here.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+on: [push]
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        ruby-version: ['2.7']
+
```
Comment (ianoxley) [inline .github/workflows/ci.yml:12]: If we're only using Ruby 2.7 then I guess we no longer need to use a matrix?

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Ruby ${{ matrix.ruby-version }}
```
Comment (ianoxley) [inline .github/workflows/ci.yml:15]:
```suggestion
      - name: Set up Ruby
```

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: ${{ matrix.ruby-version }}
```
Comment (ianoxley) [inline .github/workflows/ci.yml:18]:
```suggestion
          ruby-version: '2.7'
```

File: spec/integration/account_management_spec.rb (hunk 7 above)
Comment (banjoko01) [reply spec/integration/account_management_spec.rb:3]: The tests we are skipping are designed to rely on calling the `autheticate!` method [here](https://github.com/Sage/zuora/blob/a9af59e57518729c95f23e0c90c869005dc38e8b/lib/zuora/api.rb#L92), which then attempts to make a live network call with the correct credentials in order to return a proper auth object for the test cases we skipped. 

This was discussed during stand up last week whilst you were on vacation and we arrived at a conclusion to skip the affected tests because no one seem to have the correct credentials for the test purpose.

File: ZUORA_VERSION (hunk 6 above)
Comment (banjoko01) [reply ZUORA_VERSION:1]: I found it easier to follow the pattern used for retrieving the version in the eventq gem. Was having a bit of trouble with how the Zuora Gem version is been defined and retrieved when setting up the container and it was getting messy to get it to work when setting up the docker container (edited).

Was going to delete this [file](https://github.com/Sage/zuora/blob/master/lib/zuora/version.rb) after going with the new pattern, but missed it.

Thoughts?

File: zuora.gemspec (hunk 13 above)
Comment (banjoko01) [reply zuora.gemspec:33]: The two Gems below are the ones I introduced to resolve a couple of dependency errors encountered:
```
akami
ffi
```
The ones below already exists in the Gemfile from `master` branch, I brought them into this file in order to have them arranged in alphabetical order.
```
savon
httpi
wasabi
```

File: ZUORA_VERSION (hunk 6 above)
Comment (banjoko01) [reply ZUORA_VERSION]: Fixed in this commit [497b6d4](https://github.com/Sage/zuora/pull/14/commits/497b6d4dd7316909011bb1f950e2d4812b7046bf)

File: zuora.gemspec (hunk 13 above)
Comment (ianoxley) [reply zuora.gemspec:32]: Are these all runtime dependencies? All the development dependencies seem to have disappeared from here and are in the Gemfile instead?

You still need to remove `ruby3-backwards-compatibility`.

Also we don't want to lock gems to specific versions in the gemspec. We should be as permissive as possible with gem version specs.

**Review by ianoxley — CHANGES_REQUESTED** [review]

File: zuora.gemspec (hunk 13 above)
Comment (banjoko01) [reply zuora.gemspec:32]: `savon, httpi and wasabi` appears to be runtime dependencies, while `akami and ffi` are development dependencies. FIxing in the the PR update.

File: Gemfile [hunk 17]
```diff
@@ -1,6 +1,22 @@
-source :rubygems
+# frozen_string_literal: true
 
-gem 'wasabi', '2.3.0'
+source 'https://rubygems.org'
```
[the remark this answers is not in the document]
Comment (banjoko01) [reply Gemfile:3]: Been a public open source repo, I am not sure that would be acceptable.

@ianoxley 
What are your thoughts?

File: zuora.gemspec (hunk 13 above)
Comment (ianoxley) [reply zuora.gemspec:29]: `wasabi` should just go in the Gemfile as it's a dependency of `savon`.

File: zuora.gemspec [hunk 18]
```diff
@@ -1,36 +1,30 @@
...
-  s.homepage    = "http://github.com/wildfireapp/zuora"
-  s.summary     = %q{Zuora - ActiveModel backed client for the Zuora API}
-  s.description = %q{Zuora - Easily integrate the Zuora SOAP API using ruby objects.}
-  s.files         = `git ls-files`.split("\n")
-  s.test_files    = `git ls-files -- {test,spec,features}/*`.split("\n")
-  s.executables   = `git ls-files -- bin/*`.split("\n").map{ |f| File.basename(f) }
-  s.require_paths = ["lib"]
-  s.extra_rdoc_files = [ "README.md" ]
+  s.authors     = ['Josh Martin']
+  s.email       = ['josh.martin@wildfireapp.com']
```
Comment (ianoxley) [inline zuora.gemspec:10]: We should probably update the author and email.
```suggestion
  s.authors     = ['Sage Accounting']
  s.email       = ['sageone@sage.com']
```

File: Dockerfile [hunk 19]
```diff
@@ -0,0 +1,27 @@
+ARG RUBY_VERSION=${RUBY_VERSION:-2.7}
+FROM ruby:${RUBY_VERSION}-alpine
+
+RUN apk --update add --no-cache build-base bash && \
+  apk add git && \
+  apk add --no-cache libxml2 && \
+  apk add --no-cache libxml2-dev && \
+  apk add --no-cache sqlite-dev
+
```
Comment (ianoxley) [inline Dockerfile:9]: If we update RubyGems in the Dockerfile then there's no need to add the ffi gem to the gemspec / Gemfile:
```suggestion
RUN gem update --system 3.3.22 --no-document && gem install bundler --no-document

```

File: zuora.gemspec [hunk 20]
```diff
@@ -1,36 +1,47 @@
...
-  s.extra_rdoc_files = [ "README.md" ]
+  s.test_files  = `git ls-files -- {test,spec,features}/*`.split("\n")
+  s.extra_rdoc_files = ['README.md']
 
-  s.add_runtime_dependency('savon', ["0.9.9"])
-  s.add_runtime_dependency('activesupport', [">= 3.0.0", "< 5.0.0"])
-  s.add_runtime_dependency('activemodel', [">= 3.0.0", "< 5.0.0"])
-  s.add_runtime_dependency('httpi', ["0.9.7"])
-  s.add_runtime_dependency('libxml4r', ['~> 0.2.6'])
+  s.add_development_dependency 'akami', '<= 1.3.2'
```
Comment (ianoxley) [inline zuora.gemspec:23]: I did say this should go in the Gemfile only, as it's a dependency of `savon` (which would also make it a runtime dependency, not a development dependency). We're only using 1.3.2 as this is built against Ruby 2.7. We don't want to limit everything that uses this to akami 1.3.2 or lower.

File: zuora.gemspec [hunk 21]
```diff
@@ -1,36 +1,47 @@
...
+  s.add_development_dependency 'appraisal', '~> 0.4.1'
+  s.add_development_dependency 'artifice', '~> 0.6.0'
+  s.add_development_dependency 'bigdecimal', '< 2.0.0'
+  s.add_development_dependency 'factory_girl', '~> 2.6.4'
+  s.add_development_dependency 'ffi', '<= 1.16.3'
+  s.add_development_dependency 'guard-rspec', '~> 0.6.0'
+  s.add_development_dependency 'nokogiri', '~> 1.15.6'
```
Comment (ianoxley) [inline zuora.gemspec:30]: I did say this should go in the Gemfile only as well, as it's a dependency of `savon`.

File: zuora.gemspec [hunk 22]
```diff
@@ -1,36 +1,47 @@
...
+  s.add_development_dependency 'pry'
+  s.add_development_dependency 'pry-nav'
+  s.add_development_dependency 'rake', '~> 0.8.7'
+  s.add_development_dependency 'redcarpet', '~> 2.1.0'
+  s.add_development_dependency 'rspec', '~> 3.0'
+  s.add_development_dependency 'simplecov', '~> 0.6.4'
+  s.add_development_dependency 'sqlite3', '~> 1.3.0'
+  s.add_development_dependency 'test-unit'
+  s.add_development_dependency 'wasabi', '2.3.0'
```
Comment (ianoxley) [inline zuora.gemspec:39]: And wasabi should go in the Gemfile only too, as it is also a dependency of savon.

File: zuora.gemspec [hunk 23]
```diff
@@ -1,36 +1,47 @@
...
+  s.test_files  = `git ls-files -- {test,spec,features}/*`.split("\n")
+  s.extra_rdoc_files = ['README.md']
 
-  s.add_runtime_dependency('savon', ["0.9.9"])
-  s.add_runtime_dependency('activesupport', [">= 3.0.0", "< 5.0.0"])
-  s.add_runtime_dependency('activemodel', [">= 3.0.0", "< 5.0.0"])
-  s.add_runtime_dependency('httpi', ["0.9.7"])
-  s.add_runtime_dependency('libxml4r', ['~> 0.2.6'])
+  s.add_development_dependency 'akami', '<= 1.3.2'
+  s.add_development_dependency 'appraisal', '~> 0.4.1'
```
Comment (ianoxley) [inline zuora.gemspec:24]: We might be able to remove appraisal, as we removed those old rails3 gemfiles.

**Review by ianoxley — COMMENTED** [review]
Can you delete the `.travis.yml` file as well please?

File: zuora.gemspec (hunk 20 above)
Comment (banjoko01) [reply zuora.gemspec]: Addressed in this [commit](https://github.com/Sage/zuora/pull/14/commits/62ad549ca14815cf6660155b1f87c6399822e385)

File: zuora.gemspec (hunk 21 above)
Comment (banjoko01) [reply zuora.gemspec]: Addressed in this [commit](https://github.com/Sage/zuora/pull/14/commits/62ad549ca14815cf6660155b1f87c6399822e385)

Comment (banjoko01) [reply zuora.gemspec]: Addressed in this [commit](https://github.com/Sage/zuora/pull/14/commits/62ad549ca14815cf6660155b1f87c6399822e385)

File: zuora.gemspec (hunk 23 above)
Comment (banjoko01) [reply zuora.gemspec]: Addressed in this [commit](https://github.com/Sage/zuora/pull/14/commits/62ad549ca14815cf6660155b1f87c6399822e385)

File: zuora.gemspec (hunk 22 above)
Comment (banjoko01) [reply zuora.gemspec]: Addressed in this [commit](https://github.com/Sage/zuora/pull/14/commits/62ad549ca14815cf6660155b1f87c6399822e385)

**Review by ianoxley — APPROVED** [review]
